larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,623 @@
+	    
+	    const VertexList *pVertexList{nullptr};
+            float vtxClusterAngle{-1};
+	    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_INITIALIZED, !=, PandoraContentApi::GetList(*this,                       m_vertexListName, pVertexList));
+	    
+	    if (pVertexList && !pVertexList->empty())
+            {
+                const Vertex* pVertex{nullptr}; 
+   	        pVertex = pVertexList->front();
+	        const HitType hitType(LArClusterHelper::GetClusterHitType(cluster));
+                CartesianVector vertexPosition(LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex->GetPosition() , hitType));
+
+		const CartesianVector vtxToCluster(centroid1.GetX() - vertexPosition.GetX(), 0.f, centroid1.GetZ() - vertexPosition.GetZ());
+                const CartesianVector vtxToOtherCluster(centroid2.GetX() - vertexPosition.GetX(), 0.f, centroid2.GetZ() - vertexPosition.GetZ());
+		const float D1(vtxToCluster.GetMagnitude());
+		const float D2(vtxToOtherCluster.GetMagnitude());
+	        if ( D1*D2 != 0)
+		{
+		    vtxClusterAngle = (vtxToCluster.GetDotProduct(vtxToOtherCluster) / (D1 * D2));
+		}
+	    }
+
+	    this->EdgeHitFinder(otherCluster, otherClusterEdgeHits);
+
+	    const unsigned int otherClusterNHits{otherCluster->GetNCaloHits()};
+      	    const unsigned long int otherClusterNEdgeHits{otherClusterEdgeHits.size()};
+	    
+            CaloHitList largestList, smallestList;
+	    CartesianVector vectorTool(0.f, 0.f, 0.f);
+	    if ( clusterEdgeHits.size() > otherClusterEdgeHits.size() )
+            {
+		    largestList = clusterEdgeHits;
+		    smallestList = otherClusterEdgeHits;
+		    vectorTool = centroid2;
+            }
+	    else 
+            {
+                largestList = otherClusterEdgeHits;
+                smallestList = clusterEdgeHits;
+		vectorTool = centroid1;
+	    }
+
+            std::vector<float> distanceDistribution;
+
+	    float finalDistance{0}, maxEdgeHitSeparation{-std::numeric_limits<float>::max()}, minEdgeHitSeparation{std::numeric_limits<float>::max()}, adc1{0}, adc2{0};
+            int contact{0}, proximity{0};;
+
+	    for ( const CaloHit *const caloHit1 : largestList)
+            {
+                  CartesianVector hit1Position(caloHit1->GetPositionVector());
+                  float nearestHit{-1};
+		  adc1 += caloHit1->GetInputEnergy();
+
+		  for (const CaloHit *const caloHit2 : smallestList)
+                  {
+		      CartesianVector hit2Position(caloHit2->GetPositionVector());
+                      float distance{std::sqrt(hit1Position.GetDistanceSquared(hit2Position))};
+ 		      CartesianVector vector1(hit2Position.GetX() - hit1Position.GetX(), 0.f, hit2Position.GetZ() - hit1Position.GetZ());
+                      double smallestAngle{1000};
+                      CartesianVector tangent(centroidVector.GetZ(), 0.f, -centroidVector.GetX());
+		      const float c(tangent.GetX() * vectorTool.GetX() + tangent.GetZ() * vectorTool.GetZ());
```

Comment:
It looks like this is calculated repeatedly using the same information.  Could this be moved outside of the loop?